mllib/src/main/scala/org/apache/spark/mllib/stat/test/ChiSqTest.scala

```diff
@@ -256,4 +269,80 @@ private[spark] object ChiSqTest extends Logging {
       new ChiSqTestResult(pValue, df, statistic, methodName, NullHypothesis.independence.toString)
     }
   }
+
+   /**
+    * @param spark SparkSession
+    * @param exp A vector of expected counts for each category
+    * @param numDraw The number of monte-carlo draws to perform
+    * @param numSamplesPerPart number of samples each partition/executor will draw
+    * @param k TDigest compression factor
+    *
+    * @return A TDigest of the empirical distribution of the chi2 metric for exp
+    */
+  def getChi2Digest(
+      spark: SparkSession,
+      exp: BDV[Double],
+      numDraw: Int = 50000,
+
+      // 1073741824 = 2**30 -> ~10 minutes per digest
+      numSamplesPerPart: Int = 1073741824 / 50,
+      k: Int = 1024): TDigest = {
+
+    import breeze.linalg.{argmax, sum}
+
+    val expSum = sum(exp)
+    val expNorm = exp / expSum
+    val expFrac = { // cumulative sum
+      val e = (0 to expNorm.size).map{ i => sum(expNorm.slice(0, i)) }
+      e.slice(0, e.size - 1).zip(e.slice(1, e.size)) // boundaries for exp categories
+    }
+
+    // amount of CPU 'work' to do = constant * expSum * numDraw
+    val numPart = scala.math.max(1, (expSum * numDraw / numSamplesPerPart).ceil.toInt)
+
+    logInfo(
+      s"Starting MC simulation for chi2 digest with k=$k, numDraw=$numDraw, " +
+        s"exp.length=${exp.length}, exp=$exp, expSum=$expSum, numPart=$numPart")
+
+    val drawRange = spark.range(0, numDraw, 1, numPartitions = numPart).toDF().rdd
+
+    // Probably LOTS of room for optimization building d, but this approach works...
+    val tic = DateTime.now(DateTimeZone.UTC)
+    implicit val enc: Encoder[MergingDigest] = Encoders.kryo[MergingDigest]
+
+    val d: TDigest = drawRange.mapPartitions { part =>
+        val d: MergingDigest = new MergingDigest(k)
+        logDebug(s"building digest")
+
+        part.foreach { drawId => // do a draw
+          logDebug(s"drawId=$drawId adding to digest")
+
+          val obs = BDV.zeros[Double](exp.size)
+
+          BDV.rand(expSum.toInt).foreach { r: Double =>
+            val i = argmax(BDV(expFrac.map{ e => if (r >= e._1 && r <= e._2) 1 else 0}: _*))
+
+            obs(i) += 1
+          }
+
+          val chi2: Double = sum(breeze.numerics.pow(obs - exp, 2) / exp)
+          d.add(chi2)
+        }
+
+        logInfo(s"built digest with ${d.size()} values")
+
+        Seq(d).toIterator
+      }.treeReduce { (d1: MergingDigest, d2: MergingDigest) =>
+        logDebug(s"merging digests with ${d1.size()} values and ${d2.size()} values")
+        d1.add(d2)
+        d1
+      }
```

Comment:
Sometimes, we found the following is readable. But both has the same performance.
```scala
    val drawRange = spark.sparkContext.range(0, numDraw, 1, numPart)

    val seqOp = (d: MergingDigest, drawId: Long) => {
      val obs = BDV.zeros[Double](exp.size)
      BDV.rand(expSum.toInt).foreach { r: Double =>
        val i = argmax(BDV(expFrac.map { e => if (r >= e._1 && r <= e._2) 1 else 0 }: _*))
        obs(i) += 1
      }

      val chi2: Double = sum(breeze.numerics.pow(obs - exp, 2) / exp)
      d.add(chi2)
      d
    }

    val combOp = (d1: MergingDigest, d2: MergingDigest) => {
      d1.add(d2)
      d1
    }

    val d = drawRange.treeAggregate(new MergingDigest(k))(seqOp, combOp)

    d.compress()
    d
  }
```